Add name validation rules for different object kinds

croomes opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/master/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:
fix #9197 and #9203

**Special notes for your reviewer**:
ref
https://kubernetes.io/docs/reference/access-authn-authz/rbac/#default-roles-and-role-bindings
https://github.com/kubernetes/kubernetes/blob/585fc57cc7aaf73baf6aef84bf6dab65bde0ef9e/pkg/apis/rbac/validation/validation.go#L32

This is an alternative to https://github.com/helm/helm/pull/9218 which has some feedback but appears to have stalled.  Please feel free to reject this if you prefer to pursue that instead.

The name validation change that went into https://github.com/helm/helm/releases/tag/v3.3.0 in https://github.com/helm/helm/pull/8011 doesn't take into account the different validation rules used by the different apis.

For us, the linter is failing on the RBAC rules but I can see other Kinds are also affected.

I had a good look and I can't find any way to use a higher-level api-machinery or equivalent function to validate in a generic way - even [ValidateObjectMeta](https://pkg.go.dev/k8s.io/apimachinery@v0.20.2/pkg/api/validation#ValidateObjectMeta) needs to be passed the validation function to use.  I can't find any way to set these without comparing the object's Kind. I've not used ValidateObjectMeta directly as it checks other things besides the name.

Personally, I feel this approach is brittle but I'd like to do my part to push this closer to a resolution.  I'm not aware of the motivation behind the original PR but my vote would be to revert it.  This is the best I could come up with that matches the intended behaviour.

I've left the old validation function in the utils package as it is public.  I can remove it if requested - no Helm code is using it.

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

croomes committed:
Add name validation rules for object kinds

Signed-off-by: Simon Croome <simon@croome.org>

Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Signed-off-by: Simon Croome <simon.croome@storageos.com>

croomes commented:
@bacongobbler is this closer to what you're looking for?

adamreese commented on pkg/lint/rules/template.go:
```diff
@@ -202,18 +205,70 @@ func validateYamlContent(err error) error {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("metadata").Child("name"), obj.Metadata.Name, msg))
+	}
+	if len(allErrs) > 0 {
+		return fmt.Errorf("object name does not conform to Kubernetes naming requirements: %q (%s)", obj.Metadata.Name, allErrs.ToAggregate().Error())
```
This could be changed to use `pkg/errors`

```
return errors.Wrapf(allErrs.ToAggregate(), "object name does not conform to Kubernetes naming requirements: %q", obj.Metadata.Name)
```

croomes replied:
Good catch, fixed

adamreese commented on pkg/lint/rules/template.go:
```diff
@@ -202,18 +205,70 @@ func validateYamlContent(err error) error {
+			return apipath.IsValidPathSegmentName(name)
+		}
+	default:
+		return nil
```
Should this return `validation.NameIsDNSSubdomain` to remove the nil check in the calling function?

jdolitsky replied:
or `func(name string, prefix bool) []string { return nil }` ?

croomes replied:
I've gone with Adam's suggestion as any new object kinds are likely to use `validation.NameIsDNSSubdomain`.  Failures would be noticed immediately vs silently not validating.  Happy to take guidance on this.

adamreese reviewed (changes requested):
Thanks for doing all this research!

jdolitsky reviewed (commented):
Does this mean `chartutil.ValidateMetadataName` is to be deprecated?

croomes committed:
Move default to avoid nil check

Signed-off-by: Simon Croome <simon.croome@storageos.com>

croomes committed:
Wrap validation error instead of recreating

Signed-off-by: Simon Croome <simon.croome@storageos.com>

croomes committed:
Add/update deprecation notices

Signed-off-by: Simon Croome <simon.croome@storageos.com>

croomes commented:
> Does this mean `chartutil.ValidateMetadataName` is to be deprecated?

Yes, I think it can be.  I've added a deprecation notice and updated the previous notice.

croomes commented:
Thanks @adamreese!
